### ISSUE TITLE
Fader issue (at least with Cubase)

### ISSUE THREAD

**Issue by grizzlymv** [jean-emmanuel/open-stage-control#482]
Hi there,
     I've been experiencing a wierd issue between Open Stage Control and Cubase 9.5 Pro.  I can't find a way to have the fader (map to selected track volume) and the Knob (map to selected track panning) properly interact with Cubase. As soon as it is binded, then a bunch of random values start being added as soon as the track is selected.  I'm not sure if the problem is related to O.S.C. or Cubase, but I can't recall having the issue back when I used TouchOSC (although some gears has changed in my setup since then).  

It's fairly easy to reproduce:
- Add a Fader to O.S.C.
- MAx Value to 127. Min to 0
- Address: /Control
- Preargs (whaveter you set in Generic Remote. ex: 
[
4,
2
]
- Address to your port. ex: midi:toPCStudio
- precision can be to 0 or to its default 2, it doesn't make a difference. 

Then in Cubase:
- create a Generic Remote
- set MIDI Input to your midi port (ex: toPCStudio)
- set MIDI Output to your midi port (ex: toPCStudio)
- Add a control and name it Volume
- Set the midi channel and address to what you specified in the preargs above (ex, 4 / 2)
- set Max value to 127 (should be by default)
- Set flags to Receive and Transmit (R,T)
- in the bottom pane, set the device to VST Mixer (or Mixer)
- set the channel to Selected
- set the Value/Action to Volume
- leave the flags empty. 

Then create 2 instruments tracks
Select one and with the fader, raise the volume to around 0.  Remember the value.  
And then alternate a few time between the 2 tracks, and you'll notice that the MIDI volume will start to randomly decrease instead of staying to 0 (or whatever default value you have on your track). 

So maybe it's a bug with O.S.C.  ?  Anybody else have a similar issue with their daw?  I get the same behavior when using a knob (for track panning for instance).  Maybe I'm doing something wrong?  Every other control and feature in Cubase behave as expected.   I'm using Faders to control MIDI CC1, 2, 7, 11 and even the Pitch, and they work fine.  But they don't appear in the Generic Remote of Cubase, so I can't really do a real comparison.  

Many thanks in advance.

**Comment by grizzlymv** [comment]
And I noticed another very wierd bug.  If I set a button (in the test it was a Toggle) with a preArgs of [ something, 7] whatever the something is (1,5,8,9 doesn't matter), as long as the second one is 7, it will always send the value of the button state to the CC7.   So if your On value is 127, and Off is 0, it will max the volume of the VST when pushed (or activated if toggle), and mute it on release (or toggled).   If the off value is 60, volume will be set to half on release.   BUT, and this is where it gets interresting, it will do it even if it's not mapped in Cubase.  So what I'm saying is that the Generic Remote is NOT set (the in and out are not connected).  Despite this, the O.S.C. seems to talk to the DAW (cubase in this case) and affect the midi CC7.   If the second value of the preArgs is 1, then it would affect CC1 (Modwheel), and so on

Bottom line, O.S.C. seems to always send the second value as a Midi CC to the system, no matter if it's mapped or not, which obviously interfere with the normal operation.  

So to reproduce the problem, create a button in O.S.C., for simplicity, set is as Toggle.  address : /control  preArgs [9,1] target: midi:toyourdevice, set On to 127 and Off to 25     That's it.  Now in your daw, load an instrument track (in my case Kontakt), load a patch, then in O.S.C. press the button (modwheel should appear maxed since the second preargs value is 1) press it again (modwheel should appear close to bottom, but slightly above, because it's at 25).

**Comment by jean-emmanuel** [comment]
I don't use Cubase so this all seems a bit abstract to me, maybe there are other Cubase users here that could help. You can enable open-stage-control's `debug` option to print all sent/received messages in the console, this should help you tracking cause of the issue.

O-S-C doesn't manage the DAW's midi mapping, it just sends and receives osc/midi messages. If the messages end up controlling something they shouldn't, it means they are mapped at some point in the DAW.

> Bottom line, O.S.C. seems to always send the second value as a Midi CC to the system, no matter if it's mapped or not, which obviously interfere with the normal operation.

That's expected: the first value is the channel (http://osc.ammd.net/extras/midi/#control-channel-cc-value).

**Comment by grizzlymv** [comment]
Thx for the prompt reply Jean-Emmanuel.  What you say make sense.  However, I'm wonering why the DAW picks up the signal despite on another channel.  IE, if I send channel 9. cc 7, if my track is at midi channel 1, it should not pick up at all the cc 7.  But it does.   That did not happen when I used touchOSC.

In the example below, all I did was to press Record which have [9.7] as the preargs, and then Stop which have the 9,5 as the preargs.   You see 9,6 because that's the play button which was triggered by the DAW when I press Record.  but given by midi track is set to 1, it shouldn't pick up any of those CC values, but it still does.  This is unexpected.  Should only pick it up if I'd set the preargs to [1,7], or [1,5], etc, not [9,x]. 

In the meantime, I switched my CC value to those known as not being used (102-127) but it still get trapped in Cubase but shouldn't given the track is set to another midi channel.   

OSC sent: { address: '/control',
args: 
[ { type: 'i', value: 9 },
{ type: 'i', value: 7 },
{ type: 'i', value: 1 } ] } To : 127.0.0.1:8080
OSC received: { address: '/control', args: [ 9, 7, 1 ] } From : 127.0.0.1:8080
MIDI sent: 
OSC received: { address: '/control', args: [ 9, 7, 1 ] } From : midi:toPCStudio
MIDI received: 
OSC received: { address: '/control', args: [ 9, 5, 0 ] } From : midi:toPCStudio
MIDI received: 
OSC received: { address: '/control', args: [ 9, 6, 1 ] } From : midi:toPCStudio
MIDI received: 
OSC sent: { address: '/control',
args: 
[ { type: 'i', value: 9 },
{ type: 'i', value: 5 },
{ type: 'i', value: 1 } ] } To : 127.0.0.1:8080
MIDI sent: 
OSC received: { address: '/control', args: [ 9, 5, 1 ] } From : midi:toPCStudio
MIDI received: 
OSC received: { address: '/control', args: [ 9, 5, 1 ] } From : 127.0.0.1:8080
OSC received: { address: '/control', args: [ 9, 6, 0 ] } From : midi:toPCStudio
MIDI received: 
OSC received: { address: '/control', args: [ 9, 7, 0 ] } From : midi:toPCStudio
MIDI received:

**Comment by jean-emmanuel** [comment]
The console output you sent lacks for some informations, I suspect a windows-specific bug here I'll try to fix asap (MIDI lines are supposed to look like this: `MIDI sent: <CONTROLLER: 1 ("Modulation Wheel (coarse)"), value: 88, channel: 1>`).

Could you monitor cubase's MIDI input to check what open stage control actually sends and verify the channel is correctly set (might be another windows-specific bug) ? (https://support.native-instruments.com/hc/en-us/articles/209544729-How-to-Monitor-the-Input-of-a-MIDI-Controller)

**Comment by grizzlymv** [comment]
For the Input, I get (first is when I click on toggle button, second is when I click again the toggle button):
Timestamp, In, Port, Status, Data1, Data2, Chan, Note, Event
00006857 , 1 , --   , B8      , 6B     , 01     , 9      , --    , Control Change
00006E5D , 1 , --   , B8      , 6B     , 00     , 9      , --    , Control Change

For the Output I get:
Timestamp, In, Port, Status, Data1, Data2, Chan, Note, Event
00006857 , 1 , 3    , B8      , 6B     , 01     , 9      , --    , Control Change
00006E5D , 1 , 3    , B8      , 6B     , 00     , 9      , --    , Control Change

The toggle button has:
address: /control
preArgs: [9,107]
midi: toPCStudio

in MIDI-OX I selected toPCStudio as the Input, and my audiocard as the MIDI Output.  At first I selected toPCStudio as the out as well, but it made the loopMIDI port driver crash because of too much traffic...?  

Let me know if you want me to try anything else. 

For the record, it's the latest Cubase Pro 9.5.0.50 running on Windows 10  1803. 

Thanks

**Comment by jean-emmanuel** [comment]
This shows the MIDI messages are correctly sent on channel 9 (only), it seems cubase is tricky regarding MIDI channels : https://www.google.com/search?hl=fr&ei=CCCGXM_tK8eEjLsP0JyP8As&q=cubase+midi+control+channel&oq=cubase+midi+control+channel

**Comment by grizzlymv** [comment]
Hi there.  So, with some help, I finally figured out how to make it work properly in Cubase.  That's something others might want to know too.  So, the trick is, you'd have to create 2 MIDI ports instead of one.  One would receive all regular CCs (1, 2, 7, 10, 11, 64, etc) (let's say we name it ToPCStudioCC) and one would receive any other commands (let's name it ToPCStudio).   In Cubase, in the Studio Setup screen, under the MIDI Port Setup, you'll have to untick the in 'ALL MIDI Inputs' for the second port that will handle all other commands (ToPCStudio).  You leave the tick on for the other port (ToPCStudioCC).  Then in O.S.C. any controls for regular CC would point to the CC port you created (target Midi:ToPCStudioCC), and all other controls will points to the second port (target midi:ToPCStudio).   Then when we load O.S.C. the midi line just need to contain both ports infos and voila (--midi ToPCStudio:x,y toPCStudioCC:y,z).   A bit tricky, but work as a charm.   Thanks for pointing in the right direction.

**Comment by jean-emmanuel** [comment]
I'm glad you sorted it out !
